jax/interpreters/partial_eval.py

```diff
@@ -772,67 +773,275 @@ def _remat_partial_eval(trace, _, f, tracers, params):
+    # `concrete=True`, as an optimization we can avoid redoing *some* redundant
+    # FLOPs, namely those that produced concrete avals at the output, simply by
+    # using those as computed values. For the use case of inverse-mode ad in
+    # op-by-op ("eager mode") evaluation, all the primal outputs should be
+    # concrete (thus not recomputed).
+    to_compute = [type(pval[0]) is not ConcreteArray
+                  for uk, pval in zip(out_unknowns, eval_out_pvals) if not uk]
+    num_outputs = len(jaxpr_unknown.out_avals)
+    num_res = len(jaxpr_known.out_avals) - num_outputs
+    jaxpr_known_nores = _dce_jaxpr(jaxpr_known, out_knowns + [False] * num_res,
+                                  drop_outputs=True)
+    jaxpr_known_comp = _dce_jaxpr(jaxpr_known_nores, to_compute)
+    _, in_consts = unzip2(t.pval for t in it.chain(env_tracers, tracers))
+    reconstructed_consts = core.jaxpr_as_fun(jaxpr_known_comp)(*consts, *in_consts)
+    out_known_pvals = map(_reconstruct_pval, out_known_pvals, reconstructed_consts)
+
+    # Known outputs should keep propagating as constants
+    assert all(pv.is_known() for pv in out_known_pvals)
+    known_output_tracers = [trace.new_const(pval.get_known())
+                            for pval in out_known_pvals]
+    # Unknown outputs get wrapped in tracers with the appropriate recipe
+    unknown_output_tracers = [JaxprTracer(trace, out_pval, None)
+                              for out_pval in out_unknown_pvals]
+
+    # dce jaxpr outputs
+    new_jaxpr = _dce_jaxpr(closed_jaxpr, out_unknowns, drop_outputs=True).jaxpr
+    new_params = dict(params, call_jaxpr=new_jaxpr, differentiated=True)
+
+    # set up eqn for unknown outputs
+    const_tracers = map(trace.new_instantiated_const, consts)
+    in_tracers = (*const_tracers, *env_tracers, *instantiated_tracers)
+    eqn = new_eqn_recipe(in_tracers, unknown_output_tracers, remat_call_p,
+                         new_params, source_info_util.current())
+    for t in unknown_output_tracers: t.recipe = eqn
+    return _zip_knowns(known_output_tracers, unknown_output_tracers, out_unknowns)
 call_partial_eval_rules[remat_call_p] = _remat_partial_eval
 
 def _partition_knowns(pvals, unknowns: Sequence[bool]):
   return ([e for e, unknown in zip(pvals, unknowns) if not unknown],
           [e for e, unknown in zip(pvals, unknowns) if unknown])
 
 def _zip_knowns(known_list, unknown_list, which_unknown: Sequence[bool]):
+  assert len(known_list) + len(unknown_list) == len(which_unknown)
   known_iter, unknown_iter = iter(known_list), iter(unknown_list)
   return [next(unknown_iter) if uk else next(known_iter) for uk in which_unknown]
 
 
+def _partial_eval_jaxpr_custom(
+    jaxpr: Jaxpr, in_unknowns: List[bool], saveable: Callable[..., bool],
+  ) -> Tuple[Jaxpr, Jaxpr, List[bool], List[bool], int]:
+  if jaxpr.constvars: raise NotImplementedError  # TODO
+  env: Dict[Var, Tuple[bool, bool]] = {}
```

Comment:
perhaps use namedtuples instead of pairs of booleans? what does "inst" mean in the context of the function? we add the equations if inst is true of if it's false?